cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,673 @@
+    This is useful for converting a list of node names into a list of node names and lags.
+
+    Example:
+        ['X', 'Y', 'Z lag(n=2)'] -> [{'X': 0}, {'Y': 0}, {'Z': 2}]
+
+    :param node_names: List of node names.
+    :return: List of dictionaries with node names and lags sorted by lag and the maximum lag.
+    """
+    names_and_lags = []
+    for node_name in node_names:
+        variable_name, lag = get_variable_name_and_lag(node_name)
+        names_and_lags.append({variable_name: lag})
+    sorted_names_and_lags = sorted(names_and_lags, key=lambda x: max(x.values()), reverse=False)
+    maxlag = next(iter(sorted_names_and_lags[0].values()))
+    return sorted_names_and_lags, maxlag
+
+
+class TimeSeriesCausalGraph(CausalGraph):
+    """
+    A causal graph for time series data.
+
+    The node in a time series causal graph will have additional metadata that
+    gives the time information of the node together with the variable name.
+    The two additional metadata are:
+    - 'timedelta': the time difference with respect the reference time 0
+    - 'variable_name': the name of the variable (without the lag information).
+    """
+
+    def __init__(
+        self,
+        input_list: Optional[List[NodeLike]] = None,
+        output_list: Optional[List[NodeLike]] = None,
+        fully_connected: bool = True,
+        store_graphs: bool = False,
+    ):
+        """Initialize the time series causal graph."""
+        super().__init__(input_list, output_list, fully_connected)
+
+        self._store_graphs = store_graphs
+        # autoregressive order of the graph (max lag)
+        self._order: Optional[int] = None
+        # list of variables in the graph, i.e. discarding the lags (X1(t-1) and X1 are the same variable)
+        self._variables: Optional[List[str]] = None
+        self._summary_graph: Optional[CausalGraph] = None
+        self._minimal_graph: Optional[TimeSeriesCausalGraph] = None
+        self._meta_time_name = 'time_lag'
+        self._meta_variable_name = 'variable_name'
+
+    def __eq__(self, other: object) -> bool:
+        """Return True if the graphs are equal."""
+        if not isinstance(other, TimeSeriesCausalGraph):
+            return False
+
+        # now check if the graphs are equal. Since TimeSeriesCausalGraph is a subclass of CausalGraph,
+        # we can use the CausalGraph.__eq__ method and then check for the metadata
+        are_equal = super().__eq__(other)
+        # no need to check the metadata if the graphs are not equal
+        if not are_equal:
+            return False
+
+        # now check the metadata timedelta in the nodes
+        for node in self.get_nodes():
+            other_node = other.get_node(node.identifier)
+            # check the variable name
+            node_metadata_variable = node.get_metadata()[self._meta_variable_name]
+            other_node_metadata_td = other_node.get_metadata()[self._meta_variable_name]
```

Comment:
this would throw an error if that tag is not in the meta

Reply:
now change using the properties.